### DISCUSSION TITLE
Is there a way to convert foreign_class entity to something more production-like looking thing?

### DISCUSSION THREAD

**Discussion by k-aleksandrovich** [speechbrain/speechbrain#1665] [Q&A]
Hey guys! 👋
Here is a question. I'm using your great model from https://huggingface.co/speechbrain/emotion-recognition-wav2vec2-IEMOCAP

The proposed interface was this:
```
from speechbrain.pretrained.interfaces import foreign_class
classifier = foreign_class(source="speechbrain/emotion-recognition-wav2vec2-IEMOCAP", pymodule_file="custom_interface.py", classname="CustomEncoderWav2vec2Classifier")
out_prob, score, index, text_lab = classifier.classify_file("speechbrain/emotion-recognition-wav2vec2-IEMOCAP/anger.wav")
print(text_lab)
```

I'd like to use it in production and this raises a problem.

Each time I run the code, it loads the code and components from somewhere and puts it into generated folder `CustomEncoderWav2vec2Classifier--2392771400132274951` with a new UUID each time. It's fine while you are doing PoC, but not a great thing for a production.

Is there a way to convert the `foreign_class` entity to something more stable production-like looking thing?
Best if I could place all of the components locally and never download anything.

Cheers!

------------
MacOS Monterey v12.3

**Comment by anautsch** [comment]
Hi @k-aleksandrovich, as stated in #1032
> SpeechBrain hasn't been optimized for production.

When it comes to system integration, each production environment might have different constraints. 

[The current IEMOCAP recipe](https://github.com/speechbrain/speechbrain/tree/develop/recipes/IEMOCAP) trains and evaluates at once, as most recipes do. Are you interested in running its final step only, the `speechbrain.core.Brain.evaluate` function?

**Comment by k-aleksandrovich** [comment]
Hi @anautsch !
Thanks for the fast reply.

I'm not using IEMOCAP recipe. There is no need in training since I'm totally fine with your pretrained model. Which is why I used:
`classifier = foreign_class(source="speechbrain/emotion-recognition-wav2vec2-IEMOCAP", pymodule_file="custom_interface.py", classname="CustomEncoderWav2vec2Classifier")`
And then I just did:
`classifier.classify_file("file.wav")`


Some hack-ish workaround I came up with was this:
I took `CustomEncoderWav2vec2Classifier` and its components as-is and put it into some Wrapper like this:

 ```
 class AudioWrapper:
   def __init__(self):
        hparams_local_path = PATH_TO_AUDIO_MODEL_HYPERPARAMS
        model_components_path = PATH_TO_AUDIO_MODEL_COMPONENTS
        source = "speechbrain/emotion-recognition-wav2vec2-IEMOCAP"

        # Load the modules:
        with open(hparams_local_path) as fin:
            hparams = load_hyperpyyaml(fin, {})

        pretrainer = hparams["pretrainer"]
        pretrainer.set_collect_in(model_components_path)

        # For distributed setups, have this here:
        run_on_main(pretrainer.collect_files, kwargs={"default_source": source})

        # Load on the CPU. Later the params can be moved elsewhere by specifying run_opts={"device": ...}
        pretrainer.load_collected(device="cpu")

        self.classifier = CustomEncoderWav2vec2Classifier(modules=hparams["modules"], hparams=hparams)
        logger.info("Wav2vecClassifier initialized")
```

I copied this code from the `foreign_class`, hoping it will work. And, well, it works, but I hoped there is a way to do it in a more subtle way :)
Yet, I'm not sure if **(the stuff I just did)** will work in a stable way. What do you think?

**Comment by anautsch** [comment]
As you mentioned, there are two steps: download pre-trained models and wrapping them ready for use. This is a general concern, independent of any particular task. As you put a general `AudioWrapper`, generic wrappers for these tasks could be useful. What do you think about proposing a PR? 

We are discussing internally some related issues on data management, and this also seems to play into it. I'm wondering what the minimalistic task is - for example, [a train file](https://github.com/speechbrain/speechbrain/blob/develop/recipes/IEMOCAP/emotion_recognition/train_with_wav2vec2.py) runs the `speechbrain.core.Brain.evaluate` function at the end; in the reference train file, the calling code is:
```
    # Load the best checkpoint for evaluation
    test_stats = emo_id_brain.evaluate(
        test_set=datasets["test"],
        min_key="error_rate",
        test_loader_kwargs=hparams["dataloader_options"],
    )
```

How would you like to make this useful as for a general task (= subtle way)?
For placement, this looks more like a [template](https://github.com/speechbrain/speechbrain/tree/develop/templates) to me than a specific recipe, what are your thoughts? :)

**Comment by k-aleksandrovich** [comment]
> What do you think about proposing a PR?

I would love to, but unfortunately I have a fairly weak understanding of the underlying components of your framework. I've seen you guys have some good tutorials on the website. But sorry, I didn't have time to learn basics of your framework properly.

> speechbrain.core.Brain.evaluate

As far as I've checked, the class `CustomEncoderWav2vec2Classifier` (generated by `foreign_class`) inherits from class `Pretrained(torch.nn.Module).` So it's not quite clear what is the role of the upper-mentioned `emo_id_brain` object and 
`speechbrain.core.Brain.evaluate` in particular.

> How would you like to make this useful as for a general task (= subtle way)?

If your question refers to "what would be a good interface to use this in production", then as a typical lazy user obviously I'd prefer a Keras-like wrapper. With all the magic going on in private methods and available for me couple of public ones, like

```
wraper.load_pretrained(
model = 'your-powerful-and-clever-model-version-1.02',
path_for_components='/tmp/components'
)
```
I've no idea about used loss, hyperparams, amount of layers, etc. And I'm fine with having no idea about that :)
Once the model is here, then run:

```
wrapper.predict(
source, # df or ndarray or whatever type this model will need, I guess it'll warn me in case
)
```

Note: best if repeated launch of this "pipeline" won't download anything since `.ckpt`, `.yaml` and other files are already present.

**Comment by courentin** [comment]
@k-aleksandrovich have you found a satisfying solution?
We're running in the same kind of questioning for deploying a speechbrain model into production, and I also find the Pretrainer interface not really production adapted

**Reply by anautsch** [reply]
Hi @courentin please by all means feel free to help improve this situation; SpeechBrain is community-driven after all ;)

Please take a look at #1817 - it's meant as a starter to improve the situation, but SB's resources are limited - and for industrial needs, industry needs to get active on this platform (waiting is for the tomorrow that never comes), e.g., #1822 engages in a similar-ish discussion.
